Simple_precision_epick: an Epick-like kernel with float as FT

lrineau opened the pull request:
## Summary of Changes

New feature: if `CGAL_EPICK_SINGLE_PRECISION` is defined, then the number type of `CGAL::Epick` is `float` instead of `double`.

This pull-request also modifies the code of a few tests and examples, to that:
  - :heavy_check_mark: all of the testsuite a Kernel_23 passes,
  - :heavy_check_mark: all tests and examples of Triangulation_2 and Triangulation_3 pass,
  - :heavy_check_mark: all tests and examples of Mesh_2 pass,
  - :heavy_check_mark: all tests and examples of Mesh_3 pass.

## Rational

This pull-request is inspired by a user case, where a 3D triangulation was exported to a file format using simple-precisions floats (32 bits). After the export, with the rounding of coordinates, a few tetrahedra were inverted.

Obviously, the initial triangulation had very thin tetrahedra (or very small), because otherwise the rounding errors would not have inverted their orientations. One possible action is to try to avoid those thin/small tetrahedra. This pull-request is another approach: if the points of the triangulation have single-precision coordinates, then there would be no rounding during the exported, and those potential inverted tetrahedra would even not exist.

This "`Epick` with float" combines nice features:
  - coordinates are `float`,
  - but all predicates are evaluated using `double`:
    - static filters use `double`,
    - interval filters use `Interval_nt`, with `double`,
    - ... and then the exact type is unchanged.

A nice addition would be to wrap constructions, so that the `float` coordinates are promoted to `double` for the input of all constructions, and then the result is cast back to `float` at the end. That might be interesting for packages with constructions, and in particular for 2D constrained triangulation with intersecting constraints.

## Release Management

* Affected package(s): Kernel_23, Mesh_2, Mesh_3, Triangulation_3
* License and copyright ownership: maintenance by GeometryFactory

mglisse commented:
What is the goal?
* Is it only to reduce storage? Then it could make sense for constructions to do computations using type `double` and only convert to `float` when creating a `Point_3` or similar, although it may be harder to make it work.
* Is it also to avoid computations with `double`? Then do we need a `float` version of `Interval_nt`?

afabri commented:
The trigger is a user converting a tetrahedral mesh from double to float coordinates and experiencing inverted cells.

lrineau commented:
> What is the goal?

I just added a section "Rational" in the top message.

mglisse commented:
@lrineau not https://www.boost.org/doc/libs/1_66_0/libs/type_traits/doc/html/boost_typetraits/reference/remove_cv_ref.html ?

lrineau commented:
> @lrineau not https://www.boost.org/doc/libs/1_66_0/libs/type_traits/doc/html/boost_typetraits/reference/remove_cv_ref.html ?

You are right: that is completely equivalent. But as in CGAL we are using more and more of C++ `<type_traits>`, I prefer avoiding to pull part of `<boost/type_traits.hpp>` when that is easy.

MaelRL commented:
Should this type of STL "forwards" be in `cppXX` namespaces to make it easier to update code when the std function becomes available?

lrineau commented:
> Should this type of STL "forwards" be in `cppXX` namespaces to make it easier to update code when the std function becomes available?

Yes, probably it should be in `CGAL::cpp20::`. Please vote :+1: or :-1: so that I can see if I should really bother changing that.

lrineau commented:
@mglisse @MaelRL ... others: please review the commit 1c5454dcbe2dd0bedbca36f488b84185d59999d2, and in particular the ugly type traits, with **lots** of partial specializations, that I had to define and use to unfilter the trivial constructions (trivial constructors, and accessors to coordinates):
```c++
template <class Construction, typename ...Args>
struct Is_a_trivial_construction_base;
```

I am open to suggestions for improvements. In particular that would be nice to avoid to specialize so many templates. With C++17, we could use variable templates of type bool, in one unique class, for the same effect.

One idea of mines was to include the static function members declarations like
```c++
static Tag_true trivial(Return_base_tag, RT, RT, RT);
```
directly in the functors, and then use `has_member` and C++17 `std::is_invocable` to detect them.

But for the moment, all my possible ideas are tied to to proposal to require C++17 for CGAL-5.6.

lrineau commented:
Note that I have also added `Construct_point_on_3::operator()(Line_3)` to the concept `ConstructPointOn_3` (see commit c6858101198bb843d253388da14b3fa67183f7c0), to make `Line_3::point()` a real accessor, and not the stupid construction `point(0)`. The difference is that now `Line_3::point()` return a reference to const, instead of a constructed value.

Let's see the doc: /build:v1

lrineau commented:
> Note that I have also added `Construct_point_on_3::operator()(Line_3)` to the concept `ConstructPointOn_3` (see commit [c685810](https://github.com/CGAL/cgal/commit/c6858101198bb843d253388da14b3fa67183f7c0)), to make `Line_3::point()` a real accessor, and not the stupid construction `point(0)`. The difference is that now `Line_3::point()` return a reference to const, instead of a constructed value.

Here is the documentation I have added: [`Kernel::ConstructPointOn_3::operator()(Line_3)`](
https://cgal.github.io/6950/v1/Kernel_23/classKernel_1_1ConstructPointOn__3.html#aa18da4ad00632312e6eb56d5d9a40830).

lrineau commented on Kernel_23/include/CGAL/Exact_predicates_inexact_constructions_kernel.h:
```diff
@@ -19,28 +19,80 @@
+class Single_precision_epick;
+using Double_precision_epick = Epick;
+#else // CGAL_EPICK_SINGLE_PRECISION
+// Here, Epick is an alias to Single_precision_epick.
+class Single_precision_epick;
+class Double_precision_epick;
+using Epick = Single_precision_epick;
+#endif // CGAL_EPICK_SINGLE_PRECISION
+
+namespace internal {
+
+  // Basic objects, constructions, and predicates, using `double` for `FT`.
+  using Double_precision_epick_base =
+      Simple_cartesian<double>::Base<Double_precision_epick>::Type;
+
+  // Add the type equality property, by changing the objects types
+  using Double_precision_epick_base_with_type_equality =
+      Type_equality_wrapper<Double_precision_epick_base, Double_precision_epick>;
+
+  // Change the predicates to filtered predicates (with static filters or not)
+  using Double_precision_epick_with_filtered_predicates =
+      Filtered_kernel_adaptor<Double_precision_epick_base_with_type_equality, epick_use_static_filter>;
+
+  // Same, but with objects using `float` for FT
+  using Single_precision_epick_base =
+      Simple_cartesian<float>::Base<Single_precision_epick>::Type;
+  using Single_precision_epick_base_with_type_equality =
+      Type_equality_wrapper<Single_precision_epick_base, Single_precision_epick>;
+  using Single_precision_epick_with_filtered_predicates =
+      Filtered_kernel_adaptor<Single_precision_epick_base_with_type_equality, epick_use_static_filter>;
+};
+
+#ifndef CGAL_EPICK_SINGLE_PRECISION
+// The following is equivalent to Filtered_kernel< Simple_cartesian<double> >,
+// but it's shorter in terms of template name length (for error messages, mangling...).
+class Epick : public internal::Double_precision_epick_with_filtered_predicates
 {};
+#else // CGAL_EPICK_SINGLE_PRECISION
+class Double_precision_epick : public internal::Double_precision_epick_with_filtered_predicates
+{};
```
Please also review this new way to define `CGAL::Epick`:
  - if `CGAL_EPICK_SINGLE_PRECISION` is **not  defined** then:
    - `CGAL::Epick` is a class,
    - `CGAL::Simple_precision_epick` is a class, and
    - `CGAL::Double_precision_epick` is an alias for CGAL::Epick,
  - but if `CGAL_EPICK_SINGLE_PRECISION` **is defined** then:
    - `CGAL::Simple_precision_epick` is a class,
    - `CGAL::Double_precision_epick` is a class, and
    - `CGAL::Epick` is an alias to `CGAL::Simple_precision_epick`. That allows to see explictely that `Epick` is not really the usual one, in the compiler diagnostic messages, and in symbols names.

I have also added names and comments to the three types that were used in the previous definition of `Epick`:
```c++
class Epick
  : public Filtered_kernel_adaptor<
               Type_equality_wrapper< Simple_cartesian<double>::Base<Epick>::Type, Epick >,
               true >
{};
```

In my opinion, that is clearer that way, but you may disagree. Please comment.

lrineau commented on Kernel_23/include/CGAL/Exact_predicates_inexact_constructions_kernel.h:
```diff
@@ -19,28 +19,80 @@
-typedef Epick Exact_predicates_inexact_constructions_kernel;
+
+// This kernel is Epick with the difference that its `FT` is `float`.
+class Single_precision_epick
+    : public Converting_constructions_kernel_adaptor<internal::Single_precision_epick_with_filtered_predicates,
+                                                     Double_precision_epick>
+{};
```
Note that `Single_precision_epick` has an extra fourth layer `Converting_constructions_kernel_adaptor`, that wraps alls the constructions functors, to make then use `double` if the construction is not trivial.

MaelRL commented on Surface_mesh_parameterization/include/CGAL/Surface_mesh_parameterization/internal/validity.h:
```diff
@@ -272,11 +273,11 @@ bool is_one_to_one_mapping(const TriangleMesh& mesh,
-    b += p2.bbox();
-
-    boxes.push_back(Box(b, fd));
+    NT bx[2] = { (std::min)(p0[0], p1[0]),
+                 (std::min)(p0[1], p1[1]) };
+    NT by[2] = { (std::max)(p0[0], p1[0]),
+                 (std::max)(p0[1], p1[1]) };
+    boxes.push_back(Box(bx, by, fd));
```
p2 has been lost in the new version. Why is the change required?

Also it can just be `boxes.emplace_back(b, fd)`;

janetournois replied:
You are right, I will re-introduce `p2`.
This change is needed because `Bbox_2` hardcodes `double`s, while `Box` is `Box_with_info_d<NT>`

mglisse commented:
> the ugly type traits, with **lots** of partial specializations, that I had to define and use to unfilter the trivial constructions

From the commit message, you say that you can get infinite loops otherwise. Is that only because of the `Return_base_tag` mechanism?

lrineau commented:
> > the ugly type traits, with **lots** of partial specializations, that I had to define and use to unfilter the trivial constructions
> 
> From the commit message, you say that you can get infinite loops otherwise. Is that only because of the `Return_base_tag` mechanism?

Let's take the simple code:
```c++
CGAL::Simple_precision_epick::Point_2 p(0, 0);
```

That simple lines invokes `CGAL::Simple_precision_epick::Construct_point_2`:
https://github.com/CGAL/cgal/blob/d812473081b1614418210f80b8de0c2f0ec9ef74/Kernel_23/include/CGAL/Point_2.h#L82-L85

And thus, without any special trick, that invokes:
https://github.com/lrineau/cgal/blob/2f9e031da712e1a97dcad4946fe8130bcc67ac94/Filtered_kernel/include/CGAL/Converting_construction.h#L40-L42
```c++
auto operator()(Args&&... args) const {
    return backward_convert(construct(convert(args)...));
}
```
where `convert` and `backward_convert` are `Cartesian_convertor` objects, and `construct` is an object of type `CGAL::Epick::Construct_point_2`. The object `backward_converter` will then take a `CGAL::Epick::Point_2` object, and call:
https://github.com/CGAL/cgal/blob/d812473081b1614418210f80b8de0c2f0ec9ef74/Cartesian_kernel/include/CGAL/Cartesian_converter.h#L216-L221

And then we can back to a call of the constructor `CGAL::Simple_precision_epick::Point_2(float, float)`... and that loops. And that looping scheme was the same for a lot of constructors of kernel objects (they almost all invoke construction functors from the kernel), and in particular for all the constructors called by `Cartesian_converter<K1, K2>`.

That is why I have designed that traits class, currently named `Is_a_trivial_construction` (with a `a_` that should be removed), to "point out" which call operators are actually "trivial constructions", and can be skipped by the wrapping of `Converting_construction`. That avoids the loops with `Cartesian_converter`, and also avoids to wrap trivial constructions with unnecessary calls to `Epick` constructions.

mglisse commented:
Ah, yes, I see. I was wondering why I didn't have a similar issue in the branch NewKernel_d-Epack-glisse, but NewKernel_d is different from Kernel_23 because I would never recurse to the same Construct_point_2, the converter ends up calling the equivalent of `Simple_cartesian::Construct_point_2` (unfiltered) instead.

Knowing that a construction is trivial could probably also be used to avoid changing the rounding mode in Lazy_construction.

lrineau commented on Cartesian_kernel/include/CGAL/Cartesian/Is_a_trivial_construction.h:
```diff
@@ -0,0 +1,487 @@
+};
+
+template <class Construction, typename ...Args>
+struct Is_a_trivial_construction : public Is_a_trivial_construction_base<Construction, Args...>
```
We dealt with that discussion IRL. All seems fine.

lrineau commented on Kernel_23/include/CGAL/Exact_predicates_inexact_constructions_kernel.h:
```diff
@@ -19,28 +19,80 @@
+  using Single_precision_epick_base_with_type_equality =
+      Type_equality_wrapper<Single_precision_epick_base, Single_precision_epick>;
+  using Single_precision_epick_with_filtered_predicates =
+      Filtered_kernel_adaptor<Single_precision_epick_base_with_type_equality, epick_use_static_filter>;
```
Done: cb0ad7f5095.

lrineau commented on Kernel_23/include/CGAL/Exact_predicates_inexact_constructions_kernel.h:
```diff
@@ -19,28 +19,80 @@
+    true;
 #endif
+
+#ifndef CGAL_EPICK_SINGLE_PRECISION
+// Here Epick is a class, and Double_precision_epick an alias to it.
+class Epick;
+class Single_precision_epick;
+using Double_precision_epick = Epick;
+#else // CGAL_EPICK_SINGLE_PRECISION
+// Here, Epick is an alias to Single_precision_epick.
+class Single_precision_epick;
+class Double_precision_epick;
+using Epick = Single_precision_epick;
+#endif // CGAL_EPICK_SINGLE_PRECISION
```
We discussed that IRL as well, and we still do not completely agree.

My goal is to make sure `CGAL::Epick` does not change its behavior and semantic, unless `CGAL_EPICK_SINGLE_PRECISION`. Making `CGAL::Epick` an alias for `Double_precision_epick` would break forward declarations of it.

But, if `CGAL_EPICK_SINGLE_PRECISION` is defined, then I want to make sure that cannot be undetected. And then `CGAL::Epick` becomes an alias to `Single_precision_epick`.

MaelRL replied:
How about something like:
```
class Single_precision_epick : ... etc.
class Double_precision_epick : ... etc.
#ifdef CGAL_EPICK_SINGLE_PRECISION
class Epick : public Single_precision_epick { };
#else
class Epick : public Double_precision_epick { };
#endif

template <>
struct Triangulation_structural_filtering_traits<Epick> {
  using Use_structural_filtering_tag = Tag_true;
};
```
should be fine?

Maybe I misunderstood what you meant with "cannot be undetected".

lrineau replied:
With
```c++
class Epick : public Double_precision_epick { };
```
then `Epick` looses the "type equality property": `Epick::Point_2` and `Point_2<Epick>` would be different.

lrineau replied:
> Maybe I misunderstood what you meant with "cannot be undetected".

With the current implementation, if `CGAL_EPICK_SINGLE_PRECISION` is defined, then the debug symbols (and call stacks in compiler warnings) mention `CGAL::Simple_precision_epick` instead of `CGAL::Epick`. That helps debugging, in my opinion. It might help a user remembering that he/she has defined `CGAL_EPICK_SINGLE_PRECISION`, and it might help us if a user interacts with us about a bug.

MaelRL replied:
> With
> 
> ```c++
> class Epick : public Double_precision_epick { };
> ```
> 
> then `Epick` looses the "type equality property": `Epick::Point_2` and `Point_2<Epick>` would be different.

Right. However, one can still define a class in both cases.

I am not opposed to the current setup; I only think that the asymmetry is a cost to pay when someone has to understand that code later in the future.

lrineau replied:
Done

janetournois commented on Filtered_kernel/include/CGAL/Converting_construction.h:
```diff
@@ -0,0 +1,64 @@
+  template <typename... Args,
+            std::enable_if_t<!CartesianFunctors::Is_trivial_construction<Source_construction, Args...>::value>* = nullptr>
+  decltype(auto) operator()(Args&&... args) const {
+    return backward_convert(construct(convert(args)...));
```
This line generates a warning with visual studio `C4172 : returning address of local variable or temporary`

janetournois commented on Optimal_bounding_box/include/CGAL/Optimal_bounding_box/internal/optimize_2.h:
```diff
@@ -104,7 +104,9 @@ compute_2D_deviation(const PointRange& points,
     theta = 0.5 * CGAL_PI - theta;
 
-  return std::make_pair(pol.area(), FT{theta});
+  //cast from double to float looses data, so cast with {} is not allowed
+  //cast from double to exact types also works
```
We had decided with @lrineau to add this comment to make it clear why we are not using `{cast}`

lrineau commented on Mesh_2/include/CGAL/Mesh_2/Lipschitz_sizing_field_2.h:
```diff
@@ -144,7 +145,7 @@ class Lipschitz_sizing_field_2
   }
 
-  double operator()(const Point& p) const
+  FT operator()(const Point& p) const
```
TODO for @lrineau

lrineau replied:
`FT` is actually mandatory, because of the base class. See 36cf65d327d.

lrineau commented on Mesh_3/include/CGAL/Mesh_3/C3T3_helpers.h:
```diff
@@ -937,7 +937,7 @@ class C3T3_helpers
    */
   template <typename SliverCriterion>
-  FT min_sliver_value(const Cell_vector& cells,
+  double min_sliver_value(const Cell_vector& cells,
```
for @lrineau

lrineau replied:
Fixed in 6ef0a847b4b

lrineau commented:
All the discussions after @MaelRL's review are now resolved.

The remaining issue, in my opinion, is to find a way to test `CGAL::Single_precision_epick`, with normal CGAL packages. For the moment, it is only tested in `Kernel_23/test/`.

MaelRL commented:
> All the discussions after @MaelRL's review are now resolved.
> 
> The remaining issue, in my opinion, is to find a way to test `CGAL::Single_precision_epick`, with normal CGAL packages. For the moment, it is only tested in `Kernel_23/test/`.

Could re-introduce a macro defining the class `Epick` as `Single_precision_epick` is defined, and just define it CGAL-wide?

sloriot commented:
`remote: warning: File 6950/v1/Kernel_23/refines.html is 66.11 MB; this is larger than GitHub's recommended maximum file size of 50.00 MB`. Seems to be huge for a html file...

afabri commented:
Is it that big when you doxygen locally?

sloriot commented:
> Is it that big when you doxygen locally?

You are right, this is because of the bug of lxml/pyquery

sloriot commented:
```
    44>C:\CGAL_ROOT\CGAL-5.6-Ic-158\include\CGAL/Cartesian/Is_trivial_construction.h(62): warning C4244: 'argument': conversion from '_Ty1' to 'float', possible loss of data [C:\CGAL_ROOT\CGAL-5.6-Ic-158\cmake\platforms\MSVC2017-Release-64bits\test\Kernel_23\Filtered_cartesian.vcxproj]
                 with
                 [
                     _Ty1=int
                 ]
```
from [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.6-Ic-158/Kernel_23/TestReport_Christo_MSVC2017-Release-64bits.gz)

lrineau commented:
/build:v1

lrineau commented:
> ```
>     44>C:\CGAL_ROOT\CGAL-5.6-Ic-158\include\CGAL/Cartesian/Is_trivial_construction.h(62): warning C4244: 'argument': conversion from '_Ty1' to 'float', possible loss of data [C:\CGAL_ROOT\CGAL-5.6-Ic-158\cmake\platforms\MSVC2017-Release-64bits\test\Kernel_23\Filtered_cartesian.vcxproj]
>                  with
>                  [
>                      _Ty1=int
>                  ]
> ```
> 
> from [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.6-Ic-158/Kernel_23/TestReport_Christo_MSVC2017-Release-64bits.gz)

With `R` being "`Epick` with float`", the code triggering the warning is
```c++
  typename R::Construct_point_2 construct_point =
        rep.construct_point_2_object();
  // [...]
  Point_2 p3 = construct_point(1,1);
```
https://github.com/lrineau/cgal/blob/36cf65d327dc91843e7000320aecc80fb4f4cc74/Kernel_23/test/Kernel_23/include/CGAL/_test_new_2.h#L79-L83

The conversion `int` (literal `1`) to `float` (`FT`) triggers the warning. That will happen everywhere.

lrineau commented:
> ```
>     44>C:\CGAL_ROOT\CGAL-5.6-Ic-158\include\CGAL/Cartesian/Is_trivial_construction.h(62): warning C4244: 'argument': conversion from '_Ty1' to 'float', possible loss of data [C:\CGAL_ROOT\CGAL-5.6-Ic-158\cmake\platforms\MSVC2017-Release-64bits\test\Kernel_23\Filtered_cartesian.vcxproj]
>                  with
>                  [
>                      _Ty1=int
>                  ]
> ```
> 
> from [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.6-Ic-158/Kernel_23/TestReport_Christo_MSVC2017-Release-64bits.gz)

With `R` being "`Epick` with float`", the code triggering the warning is
```c++
  typename R::Construct_point_2 construct_point =
        rep.construct_point_2_object();
  // [...]
  Point_2 p3 = construct_point(1,1);
```
https://github.com/lrineau/cgal/blob/36cf65d327dc91843e7000320aecc80fb4f4cc74/Kernel_23/test/Kernel_23/include/CGAL/_test_new_2.h#L79-L83

The conversion `int` (literal `1`) to `float` (`FT`) triggers the warning. That will happen everywhere.

sloriot commented:
What shall we do to work around the warning? FT(1)?

afabri commented:
`1.0f` maybe

sloriot commented:
/force-build:v1

albert-github commented:
I was triggered by the remark about the size in https://github.com/CGAL/cgal/pull/6950#issuecomment-1359951223 of the page Kernel_23/refines.html

So I wanted to have a look and it looks like it is quite long , I clicked on `ComputeX_3` and landed on the page:

![image](https://user-images.githubusercontent.com/5223533/215770857-adf23ad2-3f88-4db6-bdd6-a1d7a736b956.png)

A bit strange that the bottom bar appears here in the middle, looks a bit like that some text is copied double into the html page.

Also the page Kernel_23/refines.html looks a bit strange (image from the end):

![image](https://user-images.githubusercontent.com/5223533/215773110-5f64a5ba-c1c2-4996-9ac7-d79dc3931ae5.png)


So I looked also at the pages https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-5.6-Ic-168/output2/Kernel_23/classKernel_1_1ComputeX__3.html and  https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-5.6-Ic-168/master/Kernel_23/classKernel_1_1ComputeX__3.html and they look OK (as do the corresponding Kernel_23/refines.html pages).

Question here is what causes this:
- doxygen 1.8.13, so what is the result with the current master doxygen / doxygen 1.9.6
- some post processing?

sloriot commented:
AFAIK this PR is essential for 5.6 and it is only blocked because of warnings. Do we need a meeting to decide what to do?

lrineau commented:
Please add it again to the testsuite. If that passes, it will be merged as it is.

lrineau commented:
I have merged with `master`, fixed the conflicts, and this PR passes the CI tests, now. It can be tested.

afabri commented on Installation/include/CGAL/config.h:
```diff
@@ -346,7 +346,10 @@ using std::max;
-#if __has_cpp_attribute(no_unique_address)
+#if _MSC_VER >= 1929
+// see https://devblogs.microsoft.com/cppblog/msvc-cpp20-and-the-std-cpp20-switch/#c20-no_unique_address
+#  define CGAL_NO_UNIQUE_ADDRESS [[msvc::no_unique_address]]
```
In this [testsuite](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.6-Ic-225/AABB_tree/TestReport_Christo_MSVC-2022-Community-Release.gz) we get the warning
```
include\CGAL/Lazy.h(1170,3): warning C4848: support for attribute 'msvc::no_unique_address' in C++17 and earlier is a vendor extension
``` 
Is there a fix needed, or shall we suppress the warning?

lrineau replied:
We could use:
```patch
diff --git a/Installation/include/CGAL/config.h b/Installation/include/CGAL/config.h
index c88c0c5a32c..b4b65e75e18 100644
--- a/Installation/include/CGAL/config.h
+++ b/Installation/include/CGAL/config.h
@@ -346,7 +346,7 @@ using std::max;
 #define CGAL_NORETURN  [[noreturn]]
 
 // Macro to specify [[no_unique_address]] if supported
-#if _MSC_VER >= 1929
+#if _MSC_VER >= 1929 && _MSVC_LANG >= 202002L
 // see https://devblogs.microsoft.com/cppblog/msvc-cpp20-and-the-std-cpp20-switch/#c20-no_unique_address
 #  define CGAL_NO_UNIQUE_ADDRESS [[msvc::no_unique_address]]
 #elif __has_cpp_attribute(no_unique_address)
```

sloriot commented:
Runtime error [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.6-Ic-227/Kernel_23/TestReport_lrineau_Fedora-32.gz) (in addition to the known warnings with MSVC)

sloriot commented:
Still failing [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.6-Ic-229/Kernel_23/TestReport_lrineau_Fedora-32-Release.gz)

lrineau commented:
I pushed a new commit. This time I tested it using the Docker image. It should be fine now.

afabri commented:
> * [](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.6-Ic-233/Periodic_4_hyperbolic_triangulation_2_Examples/TestReport_lrineau_Ubuntu-GCC_master_CXX20-Release.gz)
> 
> * https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.6-Ic-233/Rectangular_p_center_2_Examples/TestReport_lrineau_Ubuntu-latest-GCC6-CXX1z.gz

Should be an independent problem that we have depending on the random seed and hopefully fixed by #7407

sloriot commented:
Successfully tested in CGAL-5.6-Ic-240 (I'm doing another round of test as I was a bit aggressive with how warnings are ignored)

sloriot commented:
The warnings are still gone (but MSVC2015 but it does not really matter). Ready to be merged!
